### PR TITLE
Define an algorithm for launching a web app

### PR DESCRIPTION
**Pull request by alancutter**
This is in service of: https://github.com/WICG/web-app-launch/issues/67

This change (choose at least one, delete ones that don't apply):

* Makes editorial changes (changes informative sections, or changes normative sections without changing behavior)

Commit message:

This change adds an algorithm for launching a web app to be used by manifest features that invoke launches like shortcut items, share target, file handling etc. This is also intended to be a point for https://wicg.github.io/web-app-launch/#launching-a-web-app to replace the behaviour with something that can re-use existing web app instances.

Person merging, please make sure that commits are squashed with one of the following as a commit message prefix:

* chore:
* editorial:
* BREAKING CHANGE:
* And use none if it's a normative change

***
<a href="https://pr-preview.s3.amazonaws.com/alancutter/manifest/pull/1056.html" title="Last updated on Nov 8, 2022, 4:47 AM UTC (c81c265)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/manifest/1056/dedfbea...alancutter:c81c265.html" title="Last updated on Nov 8, 2022, 4:47 AM UTC (c81c265)">Diff</a>

### CODE REVIEW

**Comment by alancutter** [comment]
Seems like the Node CI failed with `The W3C_API_KEY input is required with VALIDATE_PUBRULES`. Maybe I need to set up something in my fork?

File: index.html
Comment (marcoscaceres) [inline index.html:2258]: Below it creates a a new browsing context, not a Service Worker window client.

File: index.html
Comment (marcoscaceres) [inline index.html:2267]:
```suggestion
          <li>Let |client| be the result of
              <a data-cite="html#creating-a-new-top-level-browsing-context">
              creating a new top-level browsing context</a>.
          </li>
```

Comment (marcoscaceres) [reply index.html:2267]: We should speak to the HTML folks about this... if we need "creating a new top-level browsing context" to be exported, we should get them to do that (and they can sanity check this).

File: index.html
Comment (marcoscaceres) [inline index.html:2270]:
```suggestion
          <li>[=Apply=] |manifest| to |client|.
          </li>
```

File: index.html
Comment (marcoscaceres) [inline index.html:2274]:
```suggestion
          <li>[=Navigate=] |client| to |url|.
          </li>
          <li>Return |client|.
          </li>
```

**Comment by marcoscaceres** [comment]
> Seems like the Node CI failed with The W3C_API_KEY input is required with VALIDATE_PUBRULES. Maybe I need to set up something in my fork?

I'll have a look. Don't worry too much about the CI errors - I'll deal with them.

File: index.html
Comment (alancutter) [reply index.html:2258]: I should update the launch_handler to use browsing context. I went with Service Worker window client in the interest of being compatible with the `clients` interface inside SW but looks like Service Worker window clients are 1:1 with navigated browsing contexts: https://w3c.github.io/ServiceWorker/#control-and-use-window-client

File: index.html
Comment (alancutter) [reply index.html:2267]: Sounds good. Filed: https://github.com/whatwg/html/issues/8449

File: index.html
Comment (marcoscaceres) [reply index.html:2258]: Ok, cool. Yeah, it's good to keep this spec SW agnostic.

File: index.html
Comment (marcoscaceres) [inline index.html:2272]: Navigating takes time... does that matter? Also, need to check if navigation can fail. 

Also, does the client need to be returned?

File: index.html
Comment (marcoscaceres) [inline index.html:81]:
```suggestion
      xref: "web-platform",
```

File: index.html
Comment (marcoscaceres) [inline index.html:86]:
```suggestion
      xref: "web-platform",
```

File: index.html
Comment (alancutter) [reply index.html]: For this particular case default navigation failure handling is sufficient.

Removed the client return.

File: index.html
Comment (alancutter) [reply index.html:2261]: They are going to export "creating a new top-level traversable" soon that this can use: https://github.com/whatwg/html/issues/8449#issuecomment-1294581251

File: index.html
Comment (alancutter) [reply index.html]: Done: https://github.com/WICG/web-app-launch/pull/81

File: index.html
Comment (alancutter) [reply index.html]: Rebased onto the new html concepts. Note that main branch now has ReSpec errors: https://github.com/w3c/manifest/issues/1061

**Comment by alancutter** [comment]
Updated with suggestions from @mgiuca.

**Comment by alancutter** [comment]
> I'll have a look. Don't worry too much about the CI errors - I'll deal with them.

Nvm I had a spec error, it's passing now.

File: index.html
Comment (mgiuca) [inline index.html:2279]: This is hand-wavy (I know we discussed it and you wanted to write this to make sure we applied the manifest without blocking on navigation) but looking closely at the **create a fresh top level traversable** algorithm, I think it's entirely valid to just apply the manifest here without having to hand-wave about the order of navigating.

- [**create a fresh top level traversable**](https://html.spec.whatwg.org/multipage/document-sequences.html#create-a-fresh-top-level-traversable) calls **navigate** automatically.
- [**navigate**](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigate) does a bunch of pre-work (should be near-instantaneous) and then "in parallel" performs the navigation. "[in parallel](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel)" is basically code for "queue a task to do this". So the **navigate** algorithm should return immediately after queueing the task, not actually blocking on completing the navigation.

This means if you apply the manifest here, it will be applied from the creation of the window, not blocked on navigation.

So just say "Apply _manifest_ to _browsing context_."

File: index.html
Comment (alancutter) [reply index.html:1400]: Move back down to the installation section and keep as non-normative?

File: index.html
Comment (mgiuca) [inline index.html:2282]: Could you sneak in a "return _browsing context_" at the end here? (And say that it returns a browsing context.)

The return value will be ignored, for now, but you'll be able to make use of it in `web-app-launch` and it's not unreasonable for this algorithm to return the browsing context it creates.

File: index.html
Comment (alancutter) [reply index.html:2252]: Done.

File: index.html
Comment (alancutter) [reply index.html:2274]: Done.

File: index.html
Comment (alancutter) [reply index.html]: Done.

File: index.html
Comment (alancutter) [reply index.html]: Done.

File: index.html
Comment (alancutter) [reply index.html:2278]: The replacement algorithm doesn't return anything either, no call site makes use of the browsing context.

Comment (alancutter) [reply index.html:2280]: Added return, it's free and could be useful for a call site one day.

**Comment by alancutter** [comment]
As discussed offline I will:
- Add prose stating the launch algorithm can be replaced by another algorithm.
- Extract creating a new client out to a separate algorithm so the replacing algorithm can still use it.

**Review by mgiuca — APPROVED** [review]
LGTM, with changes discussed offline.

**Comment by alancutter** [comment]
Requested changes made, giving @marcoscaceres a chance to review and add comments before landing.

**Pull request merged**
